File: go/kernel/verdict_test.go
// Copyright 2026 Oscar Yáñez Cisterna (@SkrOYC)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package kernel_test

import (
	"reflect"
	"testing"

	kernel "github.com/tuvren/framework/go/kernel"
)

func TestComposeVerdicts_MultipleModifiesComposeInOrder(t *testing.T) {
	first := kernel.RecordMap{
		"extension": kernel.RecordText("first"),
		"mutation":  kernel.RecordText("append-prefix"),
	}
	second := kernel.RecordMap{
		"extension": kernel.RecordText("second"),
		"mutation":  kernel.RecordText("append-suffix"),
	}

	composed := kernel.ComposeVerdicts([]kernel.Verdict{
		{Kind: kernel.VerdictKindModify, Transform: first},
		{Kind: kernel.VerdictKindProceed},
		{Kind: kernel.VerdictKindModify, Transform: second},
	})

	if composed.Kind != kernel.VerdictKindModify {
		t.Fatalf("expected modify verdict, got %s", composed.Kind)
	}

	transform, ok := composed.Transform.(kernel.RecordArray)
	if !ok {
		t.Fatalf("expected transform to be a RecordArray, got %T", composed.Transform)
	}
	if len(transform) != 2 {
		t.Fatalf("expected 2 composed transforms, got %d", len(transform))
	}
	if !reflect.DeepEqual(transform[0], kernel.Record(first)) {
		t.Errorf("first transform mismatch: %#v", transform[0])
	}
	if !reflect.DeepEqual(transform[1], kernel.Record(second)) {
		t.Errorf("second transform mismatch: %#v", transform[1])
	}
}

func TestComposeVerdicts_AbortOutranksEverything(t *testing.T) {
	abort := kernel.Verdict{Kind: kernel.VerdictKindAbort, Disposition: "HardFail", Reason: "boom"}
	composed := kernel.ComposeVerdicts([]kernel.Verdict{
		{Kind: kernel.VerdictKindModify, Transform: kernel.RecordText("ignored")},
		abort,
		{Kind: kernel.VerdictKindPause, Reason: "also ignored", ResumptionSchema: kernel.RecordNull{}},
	})

	if composed.Kind != kernel.VerdictKindAbort || composed.Reason != "boom" {
		t.Fatalf("expected the abort verdict to win, got %+v", composed)
	}
}

func TestComposeVerdicts_SingleModifyStaysUnwrapped(t *testing.T) {
	only := kernel.RecordText("solo")
	composed := kernel.ComposeVerdicts([]kernel.Verdict{
		{Kind: kernel.VerdictKindProceed},
		{Kind: kernel.VerdictKindModify, Transform: only},
	})

	if composed.Kind != kernel.VerdictKindModify {
		t.Fatalf("expected modify verdict, got %s", composed.Kind)
	}
	if !reflect.DeepEqual(composed.Transform, kernel.Record(only)) {
		t.Errorf("expected single modify transform to stay unwrapped, got %#v", composed.Transform)
	}
}

func TestComposeVerdicts_EmptyOrAllProceedYieldsProceed(t *testing.T) {
	if composed := kernel.ComposeVerdicts(nil); composed.Kind != kernel.VerdictKindProceed {
		t.Fatalf("expected proceed for empty input, got %s", composed.Kind)
	}
	composed := kernel.ComposeVerdicts([]kernel.Verdict{
		{Kind: kernel.VerdictKindProceed},
		{Kind: kernel.VerdictKindProceed},
	})
	if composed.Kind != kernel.VerdictKindProceed {
		t.Fatalf("expected proceed, got %s", composed.Kind)
	}
}

File: dart/kernel/test/verdict_test.dart
// Copyright 2026 Oscar Yáñez Cisterna (@SkrOYC)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:test/test.dart';
import 'package:tuvren_kernel/tuvren_kernel.dart';

void main() {
  test('multiple modifies compose in registration order', () {
    final first = RecordMap({
      'extension': const RecordText('first'),
      'mutation': const RecordText('append-prefix'),
    });
    final second = RecordMap({
      'extension': const RecordText('second'),
      'mutation': const RecordText('append-suffix'),
    });

    final composed = composeVerdicts([
      Verdict(kind: verdictKindModify, transform: first),
      const Verdict(kind: verdictKindProceed),
      Verdict(kind: verdictKindModify, transform: second),
    ]);

    expect(composed.kind, verdictKindModify);
    final transform = composed.transform;
    expect(transform, isA<RecordArray>());
    final elements = (transform as RecordArray).value;
    expect(elements, hasLength(2));
    expect(elements[0], equals(first));
    expect(elements[1], equals(second));
  });

  test('abort outranks everything', () {
    const abort = Verdict(
      kind: verdictKindAbort,
      disposition: 'HardFail',
      reason: 'boom',
    );
    final composed = composeVerdicts([
      const Verdict(kind: verdictKindModify, transform: RecordText('ignored')),
      abort,
      const Verdict(
        kind: verdictKindPause,
        reason: 'also ignored',
        resumptionSchema: RecordNull(),
      ),
    ]);

    expect(composed.kind, verdictKindAbort);
    expect(composed.reason, 'boom');
  });

  test('single modify stays unwrapped (not an array)', () {
    const only = RecordText('solo');
    final composed = composeVerdicts([
      const Verdict(kind: verdictKindProceed),
      const Verdict(kind: verdictKindModify, transform: only),
    ]);

    expect(composed.kind, verdictKindModify);
    expect(composed.transform, equals(only));
  });

  test('empty or all-proceed input yields proceed', () {
    expect(composeVerdicts(const []).kind, verdictKindProceed);

    final composed = composeVerdicts(const [
      Verdict(kind: verdictKindProceed),
      Verdict(kind: verdictKindProceed),
    ]);
    expect(composed.kind, verdictKindProceed);
  });

  test('rejects unknown verdict kind', () {
    expect(
      () => composeVerdicts(const [Verdict(kind: 'not-a-real-verdict-kind')]),
      throwsFormatException,
    );
  });

  test('pause outranks modify, retry, and proceed', () {
    final composed = composeVerdicts([
      const Verdict(kind: verdictKindRetry, adjustment: RecordInt(1)),
      const Verdict(kind: verdictKindModify, transform: RecordText('x')),
      const Verdict(kind: verdictKindProceed),
      const Verdict(
        kind: verdictKindPause,
        reason: 'need approval',
        resumptionSchema: RecordText('z'),
      ),
    ]);
    expect(composed.kind, verdictKindPause);
    expect(composed.reason, 'need approval');
  });

  test('retry outranks proceed', () {
    final composed = composeVerdicts(const [
      Verdict(kind: verdictKindProceed),
      Verdict(kind: verdictKindRetry, adjustment: RecordInt(3)),
    ]);
    expect(composed.kind, verdictKindRetry);
    expect(composed.adjustment, equals(const RecordInt(3)));
  });

  test('a single transform-less modify composes as RecordNull, not a throw',
      () {
    // Regression: go/kernel/verdict.go's Transform field is a plain Record
    // interface value, so a nil transform composes without error there;
    // this pins the same tolerance for Dart's nullable transform field.
    final composed = composeVerdicts(const [
      Verdict(kind: verdictKindModify),
    ]);

    expect(composed.kind, verdictKindModify);
    expect(composed.transform, equals(const RecordNull()));
  });

  test(
      'a transform-less modify composes alongside other modifies without '
      'throwing', () {
    const withTransform = RecordText('has-transform');
    final composed = composeVerdicts(const [
      Verdict(kind: verdictKindModify, transform: withTransform),
      Verdict(kind: verdictKindModify),
    ]);

    expect(composed.kind, verdictKindModify);
    final transform = composed.transform;
    expect(transform, isA<RecordArray>());
    final elements = (transform as RecordArray).value;
    expect(elements, hasLength(2));
    expect(elements[0], equals(withTransform));
    expect(elements[1], equals(const RecordNull()));
  });
}
